File: 0149-max-points-on-a-line/0149-max-points-on-a-line.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        from math import gcd
        
        def getSlope(p1, p2):
            dx = p2[0] - p1[0]
            dy = p2[1] - p1[1]
            if dx == 0: return (0, 1)
            if dy == 0: return (1, 0)
            g = gcd(abs(dx), abs(dy))
            if dx < 0: dx, dy = -dx, -dy
            return (dx // g, dy // g)
        
        n = len(points)
        if n <= 2: return n
        
        res = 2
        for i in range(n):
            slopes = defaultdict(int)
            for j in range(i+1, n):
                slope = getSlope(points[i], points[j])
                slopes[slope] += 1
                res = max(res, slopes[slope] + 1)
        
        return res

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.cpp
class Solution {
public:
    int maxPoints(vector<vector<int>>& points) {
        int n = points.size();
        if (n <= 2) return n;
        int result = 2;
        
        for (int i = 0; i < n; i++) {
            map<pair<int,int>, int> slopes;
            
            for (int j = i + 1; j < n; j++) {
                int dx = points[j][0] - points[i][0];
                int dy = points[j][1] - points[i][1];
                int g = __gcd(abs(dx), abs(dy));
                dx /= g; dy /= g;
                
                if (dx < 0) { dx = -dx; dy = -dy; }
                else if (dx == 0) dy = abs(dy);
                
                slopes[{dx, dy}]++;
                result = max(result, slopes[{dx, dy}] + 1);
            }
        }
        
        return result;
    }
};